### ISSUE TITLE
More constructor options for Ort::Session

### ISSUE THREAD

**Issue by BillMcCrary** [microsoft/onnxruntime#4131]
**Is your feature request related to a problem? Please describe.**
We want to be able to declare an Ort::Session without knowing the file path of the model.

**System information**
ORT 1.2, C++ (VC++ 2019), Win10

**Describe the solution you'd like**
We would like to be able to do this in our class header:
`Session ortSession;`

And then later, elsewhere in our class definition, we'd like to be able to set this file dynamically:
`ortSession = Session(ortEnv, modelPath, SessionOptions{ nullptr });`

Right now, in our header, we have to pass something for the file, but at runtime as an installed windows service this isn't known when we create an instance.

**Describe alternatives you've considered**
Right now the brute force option is to hard code our ONNX file paths, like this:
`Session ortSession{ ortEnv, L"c:/ONNX/CameraViewModelMultiClass.onnx", SessionOptions{ nullptr } };`

Since this is running as a windows service, setting a relative path here is unclear right now, still chewing on options for this.

**Additional context**
We want to create this session 1x and then infer over and over (vision inspection system).  We COULD create the session on every inference pass but that affects performance.

This is possibly in some documentation or example somewhere but I have yet to find it...

**Comment by hariharans29** [comment]
Can you just point to a session and when you know the model path instantiate the session ? Right now, this would be hard to achieve given that session construction does a lot of things under the hood with the given model and a constructed session without an associated model isn't much really.

"We want to create this session 1x and then infer over and over (vision inspection system)."

You can do this for sure. A session needs to be constructed only once. You can call `Run()` as many times as you want to score the model associated with the session object.

**Comment by BillMcCrary** [comment]
Well I remember having tried that before with some errors.  I tried again just now, and hopefully I'm doing something wrong - maybe you can take a look at these snippets and please correct me?  I should disclaim my C++ is a little rusty:

- In my header where I declare both a pointer and an instance:
`Session* pOrtSession;`
`Session ortSession{ ortEnv, L"C:/FullPathToSome.onnx", SessionOptions{ nullptr } };`

- Here is where I later dynamically change the session model:
`Session ortSession2 = Session(ortEnv, modelPath, SessionOptions{ nullptr });
pOrtSession = &ortSession2;`
`ortSession = Session(ortEnv, modelPath, SessionOptions{ nullptr }); `

- Then when I try to call Run, the instance succeeds but the pointer throws an error:
`ortSession.Run(...)`
`pOrtSession->Run(...)`

- Error snapshot
![image](https://user-images.githubusercontent.com/5295452/84053018-5463f880-a97f-11ea-9428-6d1f1ec774dc.png)

**Comment by skottmckay** [comment]
What's the reason the class can't have a std::unique_ptr<Session> member with you allocating that when you have the necessary info like the model path, and using that for the Run calls?

As for why you get an error, without seeing the full code it's hard to say. pOrtSession was set to the address of ortSession2, but it's not clear if ortSession2 is still in scope when you attempted the pOrtSession->Run call. If it is not, the session it pointed to has been freed.

**Comment by BillMcCrary** [comment]
@hariharans29 & @skottmckay thanks!  That was the issue, working for me now.  It's been 15 years since I used C++ so I'm relearning a lot as I go - I had seen but didn't quite understand the unique_ptr ownership rules until now.  

For anyone who sees this, here is the final working bits:

Header:
```
unique_ptr<Session> pOrtSession = NULL;
```

Implementaion:
```
auto uOrtSession = make_unique<Session>(ortEnv, modelPath, SessionOptions{ nullptr });
pOrtSession = move(uOrtSession);
```

**Comment by hariharans29** [comment]
Thanks for letting us know ! :)

**Comment by hcfeng201** [comment]
> @hariharans29 & @skottmckay thanks! That was the issue, working for me now. It's been 15 years since I used C++ so I'm relearning a lot as I go - I had seen but didn't quite understand the unique_ptr ownership rules until now.
> 
> For anyone who sees this, here is the final working bits:
> 
> Header:
> 
> ```
> unique_ptr<Session> pOrtSession = NULL;
> ```
> 
> Implementaion:
> 
> ```
> auto uOrtSession = make_unique<Session>(ortEnv, modelPath, SessionOptions{ nullptr });
> pOrtSession = move(uOrtSession);
> ```

Hi,
    I try to implement it like you did，but got en error.
header:
    ```std::unique_ptr<Ort::Session> onnx_session_ = NULL;
      Ort::Env *env_;```
Implementaion:
``` 
bool loadONNX() {
        const char* onnx_model_path_ = "model/melgan_dynamic.onnx";
        Ort::Env env_(ORT_LOGGING_LEVEL_ERROR, "test");
        Ort::SessionOptions session_options;
        session_options.SetIntraOpNumThreads(1);
        session_options.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_EXTENDED);
        auto uOrtSession = std::make_unique<Ort::Session>(env_, onnx_model_path_, session_options);
        onnx_session_ = std::move(uOrtSession);
        return ture;
}
void RunONNXSession(std::vector<float> &onnx_input_tensor, std::vector<int64_t> &onnx_input_node_dims_, short 
                               *body_data, int64_t &out_size) {
        const char* input_names[] = {"input"};
        const char* output_names[] = {"output"};
        auto memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
        size_t input_tensor_size = onnx_input_node_dims_[0] * onnx_input_node_dims_[1] * onnx_input_node_dims_[2];
        Ort::Value input_tensor = Ort::Value::CreateTensor<float>(memory_info, onnx_input_tensor.data(), input_tensor_size, 
                                                              onnx_input_node_dims_.data(), 3);
        auto output_tensors = onnx_session_->Run(Ort::RunOptions{nullptr}, input_names, &input_tensor, 1, output_names, 1);
        ...
} 
```
    
    
![image](https://user-images.githubusercontent.com/34930362/87645042-8424ce00-c77f-11ea-9f55-5d50c161afd8.png)

**Comment by goksinan** [comment]
If I use a pointer, I still get the following error when I do `session->Run(...) `

```
[E:onnxruntime:, sequential_executor.cc:309 onnxruntime::SequentialExecutor::Execute] Non-zero status code returned while running Conv node. Name:'Conv_0' Status Message: Attempt to use DefaultLogger but none has been registered.
```

However, I don't get an error if I load and do `session.Run(...)` inside the same function.
I would appreciate any feedback.

**Comment by arbasli2** [comment]
> Hi,
> I try to implement it like you did，but got en error.
> header:
> 
> ```
> std::unique_ptr<Ort::Session> onnx_session_ = NULL;
> Ort::Env *env_;
> ```
> 
> Implementaion:
> 
> ```
> bool loadONNX() {
>         const char* onnx_model_path_ = "model/melgan_dynamic.onnx";
>         Ort::Env env_(ORT_LOGGING_LEVEL_ERROR, "test");

It seems since you have used `env` variable to create the session when it goes out of scope, it creates error when you use the session. so for me it worked when I did  the same trick to the `env`  as is suggested for `session` variable, i.e. defined it as a unique pointer and then moved the local `env` to this class member:

in header: 

    std::unique_ptr<Ort::Session> session= nullptr;
    std::unique_ptr < Ort::Env> env = nullptr;

in the constructor of my class:

    Ort::SessionOptions session_options{nullptr};
    auto envLocal = std::make_unique<Ort::Env>(ORT_LOGGING_LEVEL_WARNING, "test");
    env = std::move(envLocal);
    auto sessionLocal = std::make_unique<Ort::Session>(*env, model_path, session_options);
    session = std::move(sessionLocal);

**Comment by goksinan** [comment]
I just figured out the same thing two days ago :) As you said, when the initialization function goes out of scope, `session->run()` gives an error when you call it from another member function. But in my case, I also had to run the model once right after initialization to eliminate all the errors. In other words, I created the session, also created a dummy input, and ran the session using this dummy input (all inside the same function).

Interestingly, the above trick (using unique pointer for Ort::Env) is not necessary if you download and use a model from the Onnx model zoo. I tested `squeezenet.onnx` and it worked just fine. So, I am still thinking that there might be a bug that causes all of this.

**Comment by yuminhhuang** [comment]
> > Hi,
> > I try to implement it like you did，but got en error.
> > header:
> > ```
> > std::unique_ptr<Ort::Session> onnx_session_ = NULL;
> > Ort::Env *env_;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Implementaion:
> > ```
> > bool loadONNX() {
> >         const char* onnx_model_path_ = "model/melgan_dynamic.onnx";
> >         Ort::Env env_(ORT_LOGGING_LEVEL_ERROR, "test");
> > ```
> 
> It seems since you have used `env` variable to create the session when it goes out of scope, it creates error when you use the session. so for me it worked when I did the same trick to the `env` as is suggested for `session` variable, i.e. defined it as a unique pointer and then moved the local `env` to this class member:
> 
> in header:
> 
> ```
> std::unique_ptr<Ort::Session> session= nullptr;
> std::unique_ptr < Ort::Env> env = nullptr;
> ```
> 
> in the constructor of my class:
> 
> ```
> Ort::SessionOptions session_options{nullptr};
> auto envLocal = std::make_unique<Ort::Env>(ORT_LOGGING_LEVEL_WARNING, "test");
> env = std::move(envLocal);
> auto sessionLocal = std::make_unique<Ort::Session>(*env, model_path, session_options);
> session = std::move(sessionLocal);
> ```

This comment works for me. Must `env` be moved in addition to the `session` were moved.
